### PR TITLE
BLOCKCHAIN-439 Congress schedule spend all assets and remark them

### PR DESCRIPTION
**Pull request by MichalLustykNeti**
Depends on https://github.com/liberland/liberland_substrate/pull/380

### CODE REVIEW

**Comment by kacperzuk-neti** [comment]
@MichalLustykNeti #380 merged

**Comment by kacperzuk-neti** [comment]
https://github.com/liberland/liberland_substrate/pull/380 is merged

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]
This PR title is "Congress schedule spend all assets and remark them #287". But I don't see neither scheduling nor remarking anywhere in here. Did I miss it or is it not implemented?

File: yarn.lock
Comment (kacperzuk-neti) [inline yarn.lock:1]: We have no changes to package.json, there shouldn't be any yarn.lock changes.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by SzymonZiebaNeti — APPROVED** [review]

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1362,29 +1362,68 @@ const createProposalAndVote = async (threshold, proposalContent, vote) => {
...
   const threshold = await congressMajorityThreshold();
-
-  const executeProposal = api.tx.llm.sendLlm(transferToAddress, transferAmount);
+  const remark = api.tx.llm.remark('Congress spending');
+  const transferAndRemark = api.tx.utility.batchAll([spendProposal, remark]);
+  const dayToBlocks = (24 * 60 * 60) / 6;
+  const motionDuration = 7 * dayToBlocks;
+
+  // CouncilAccount execute forces Congress to use scheduler with at least 4 days delay.
+  // We add 1 extra day in case of any desync between signing and sending tx.
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1375]: There's no `desync`, there's nothing that kept in sync.
The additional 1 day is for some person to actually Close (a.k.a. execute) the motion after it ends, which is done manually.

File: src/api/nodeRpcCall.js [hunk 2]
```diff
@@ -1362,29 +1362,68 @@ const createProposalAndVote = async (threshold, proposalContent, vote) => {
...
+  const transferAtBlock = blockNumber + (5 * dayToBlocks) + motionDuration;
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1376]: Just a note, as we may want to make this smarter in some future PR.

So in this PR it's hardcoded:
1. 7 days for motion duration
2. 1 day for someone to execute the motion after voting ends
3. 4 days for Senate to potentially veto

That's for the worst-case scenario that we didn't reach the threshold of signatures before motion ended and the vote of prime member must be used to fill for non-voters.
However, if congress coordinates and gets the threshold sooner, we only need 4 days + the time needed for congress to coordinate, which could be as low as an hour or a day.
The UX for that is hard to get right though. Maybe we could allow proposer to specify the delay manually and only enforce 4d minimum?

File: src/api/nodeRpcCall.js [hunk 3]
```diff
@@ -1362,29 +1362,68 @@ const createProposalAndVote = async (threshold, proposalContent, vote) => {
...
 };
 
-const congressSendLlm = async ({ walletAddress, transferToAddress, transferAmount }) => {
+const congressProposeSpend = async ({ walletAddress, blockNumber, spendProposal }) => {
   const api = await getApi();
 
   const threshold = await congressMajorityThreshold();
-
-  const executeProposal = api.tx.llm.sendLlm(transferToAddress, transferAmount);
+  const remark = api.tx.llm.remark('Congress spending');
```
[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js:1369]: Oh, and important thing: this has 64 characters limit, we should enforce it client-side.

File: src/api/nodeRpcCall.js [hunk 4]
```diff
@@ -1362,29 +1362,68 @@ const createProposalAndVote = async (threshold, proposalContent, vote) => {
...
+  const transferAndRemark = api.tx.utility.batchAll([spendProposal, remark]);
+  const dayToBlocks = (24 * 60 * 60) / 6;
+  const motionDuration = 7 * dayToBlocks;
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/api/nodeRpcCall.js:1372]: image
Motion duration is no const. Do you have any idea how we could query that?

[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js:1374]: Huh, I was sure it is a constant. In that case, we indeed need to leave it hardcoded.

File: src/components/Modals/SpendAssetModal.js [hunk 5]
```diff
@@ -64,16 +72,51 @@ function SpendAssetModal({ closeModal, assetData }) {
...
+        placeholder={`Amount ${assetData.metadata.name} (${assetData.metadata.symbol})`}
         required
       />
       { errors?.amount?.message
         && <div className={styles.error}>{errors.amount.message}</div> }
 
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        validate={(textDescription) => Validator.validateLen(3, textDescription)}
```
Comment (kacperzuk-neti) [inline src/components/Modals/SpendAssetModal.js:96]: We should enforce 196 chars max here.

File: src/components/Modals/SpendLLDModal.js [hunk 6]
```diff
@@ -99,6 +91,33 @@ function SpendLLDModal({ closeModal }) {
         <div className={styles.error}>{errors.amount.message}</div>
       )}
 
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        validate={(textDescription) => Validator.validateLen(3, textDescription)}
```
Comment (kacperzuk-neti) [inline src/components/Modals/SpendLLDModal.js:97]: We should enforce 196 chars max here.

File: src/components/Modals/SpendLLMModal.js [hunk 7]
```diff
@@ -75,14 +73,41 @@ function SpendLLMModal({
...
         placeholder="Amount LLM"
         required
       />
       { errors?.amount?.message
         && <div className={styles.error}>{errors.amount.message}</div> }
 
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        validate={(textDescription) => Validator.validateLen(3, textDescription)}
```
Comment (kacperzuk-neti) [inline src/components/Modals/SpendLLMModal.js:87]: We should enforce 196 chars max here.

File: src/components/Modals/SpendPolitipoolLLMModal.js [hunk 8]
```diff
@@ -75,14 +73,41 @@ function SpendPolitipoolLLMModal({
...
         placeholder="Amount LLM"
         required
       />
       { errors?.amount?.message
         && <div className={styles.error}>{errors.amount.message}</div> }
 
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        validate={(textDescription) => Validator.validateLen(3, textDescription)}
```
Comment (kacperzuk-neti) [inline src/components/Modals/SpendPolitipoolLLMModal.js:87]: We should enforce 196 chars max here.

File: src/components/Modals/SpendAssetModal.js [hunk 9]
```diff
@@ -64,16 +72,51 @@ function SpendAssetModal({ closeModal, assetData }) {
...
+        name="description"
+        placeholder="Spend description"
+        required
+      />
+      { errors?.description?.message
+        && <div className={styles.error}>{errors.description.message}</div> }
+
+      <div className={styles.title}>Senate voting duration in days</div>
```
Comment (kacperzuk-neti) [inline src/components/Modals/SpendAssetModal.js:104]: I don't understand, why are we allowing user to change this? This is always 4 days.

What could be adjusted is the time allocated for getting this through the Congress, as this could be as short as 1 day if congress coordinates.

The 4 days is a blockchain requirement - we need 4 days between the congress motion gets executed and the time scheduler should execute the call. User can't decide that.

It it just a confusion of Congress/Senate and this should be `Congress voting duration in days`? But in the nodeRpcCall it indeed is put in the place that refers to the time that Senate has for veto, not motionDuration.

Can you describe what was the intent here?

File: src/utils/validator.js [hunk 10]
```diff
@@ -10,4 +13,19 @@ export default class Validator {
...
+    try {
+      const unbondValue = parseMerits(textUnbondValue);
+      if (unbondValue.gt(maxUnbond) || unbondValue.lte(BN_ZERO)) return 'Invalid amount';
+      return true;
+    } catch (e) {
+      return 'Invalid amount';
+    }
+  }
+
+  static validateLen(len, text) {
```
Comment (kacperzuk-neti) [inline src/utils/validator.js:27]: Why not simply use `minLength` and `maxLength` instead? https://react-hook-form.com/docs/useform/register

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: yarn.lock
Comment (kacperzuk-neti) [inline yarn.lock:1]: yarn.lock shouldn't change if package.json doesn't.

File: src/api/nodeRpcCall.js [hunk 11]
```diff
@@ -13,7 +13,7 @@ import { convertAssetData } from '../utils/dexFormatter';
 import { parseDollars, parseMerits } from '../utils/walletHelpers';
 import { getMetadataCache, setMetadataCache } from '../utils/nodeRpcCall';
 import { addReturns, calcInflation, getBaseInfo } from '../utils/staking';
-import identityJudgementEnums from "../constants/identityJudgementEnums";
+import identityJudgementEnums from '../constants/identityJudgementEnums';
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:16]: Not directly related to the PR, but pre-commit hook wasn't allowing me to commit without fixing this.

File: src/components/Home/HomeNavigation/index.js [hunk 12]
```diff
@@ -211,7 +211,7 @@ function HomeNavigation() {
                     icon={icon}
                     path={location.pathname}
                     description={description}
-                    isDiscouraged={isDiscouraged}
+                    isDiscouraged={isDiscouraged === 'true'}
```
Comment (kacperzuk-neti) [inline src/components/Home/HomeNavigation/index.js:214]: @SzymonZiebaNeti can you comment on this change by Michal? I'm not sure how is this related to this PR.

**Review by kacperzuk-neti — APPROVED** [review]

**Review by SzymonZiebaNeti — APPROVED** [review]

File: src/components/Modals/SpendAssetModal.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        required
+      />
+      { errors?.amount?.message
+        && <div className={styles.error}>{errors.amount.message}</div> }
+
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        minLength={{ value: 3, message: 'Min description length is 3 chars' }}
+        maxLength={{ value: 256, message: 'Min description length is 256 chars' }}
```
Comment (DorianSternVukotic) [inline src/components/Modals/SpendAssetModal.js:98]: max instead of min

File: src/components/Modals/SpendLLDModal.js [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        <div className={styles.error}>{errors.amount.message}</div>
+      )}
+
+      <div className={styles.title}>Spend description</div>
+      <TextInput
+        register={register}
+        minLength={{ value: 3, message: 'Min description length is 3 chars' }}
+        maxLength={{ value: 256, message: 'Min description length is 256 chars' }}
+        name="description"
+        placeholder="Spend description"
```
Comment (DorianSternVukotic) [inline src/components/Modals/SpendLLDModal.js:100]: why are spendLLMModal and spendAssetModal different modals? 

I understand for spendPolitipooledLLMModal, but thats different
